src/lib/updating-element.ts

```diff
@@ -0,0 +1,503 @@
+ * Uses `getProperty` and `setProperty` to manage the property's value.
+ */
+const makeProperty = (name: string, proto: Object) => {
+  if (proto.hasOwnProperty(name) || (name in proto)) {
+    return;
+  }
+  Object.defineProperty(proto, name, {
+    get() {
+      return this.getProperty(name);
+    },
+    set(value) {
+      this.setProperty(name, value);
+    },
+    configurable: true,
+    enumerable: true
+  });
+};
+
+/**
+ * Creates and sets object used to memoize all class property values. Object
+ * is chained from superclass.
+ */
+const ensurePropertyStorage = (ctor: typeof UpdatingElement) => {
+  if (!ctor.hasOwnProperty('_classProperties')) {
+    ctor._classProperties = Object.create(Object.getPrototypeOf(ctor)._classProperties);
+  }
+};
+
+/**
+ * Decorator which creates a property. Optionally a `PropertyOptions` object
+ * can be supplied to describe how the property should be configured.
+ */
+export const property = (options?: PropertyOptions) => (proto: Object, name: string) => {
+  const ctor = proto.constructor as typeof UpdatingElement;
+  ensurePropertyStorage(ctor);
+  ctor._classProperties[name] = options || {};
+  makeProperty(name, proto);
+};
+
+
+/**
+ * AttributeProcessor which configures properties which should reflect to and
+ * from boolean attributes. If the attribute exists, the property becomes true;
+ * and if the property value is truthy, the attribute is set to an empty string.
+ */
+export const BooleanAttribute: AttributeProcessor = {
+  fromAttribute: (value: string) => value !== null,
+  toAttribute: (value: string) => value ? '' : null
+};
+
+/**
+ * Change function that returns true if `value` is different from `oldValue`.
+ * This method is used as the default for a property's `shouldChange` function.
+ */
+// tslint:disable-next-line no-any
+export const identity = (value: any, old: any) => {
+  // This ensures (old==NaN, value==NaN) always returns false
+  return old !== value && (old === old || value === value);
+};
+
+enum ValidationState {
+  Disabled = 0,
+  Valid,
+  Invalid
+}
+
+/**
+ * Base element class which manages element properties and attributes. When
+ * properties change, the `update` method is asynchronously called. This method
+ * should be supplied by subclassers to render updates as desired.
+ */
+export abstract class UpdatingElement extends HTMLElement {
+
+  /**
+   * Maps attribute names to properties; for example `foobar` attribute
+   * to `fooBar` property.
+   */
+  private static _attributeToPropertyMap: AttributeMap = {};
+  /**
+   * Marks class as having finished creating properties.
+   */
+  private static _finalized = true;
+  /**
+   * Memoized result of computed observedAttributes.
+   */
+  private static _observedAttributes: string[]|undefined;
+
+  // TODO(sorvell): intended to be private but called by decorator.
+  /**
+   * Memoized list of all class properties, including any superclass properties.
+   */
+  static _classProperties: Properties = {};
```

Comment:
Why do you need this and `properties`? Can't the decorator just write to `properties`?

Reply:
It's a memoization for efficiency so we don't need to look at the superclass properties at runtime. This is info is really frequently needed: when a property invalidates or deserializes, for example.